### ISSUE TITLE
Calling `Conn.Close()` on a blocking call doesn't interrupt the call.

### ISSUE THREAD

**Issue by ereOn** [garyburd/redigo#239]
Hi,

I noticed a surprising behavior of `Conn.Close()` when the connection is in a pending blocked call.

Please consider the following scenario:

```
conn := redisPool.Get()
defer conn.Close()

go func() {
    // Blocking call. We block forever (0 timeout).
    redisClient.Do("BRPOP", keyQueue, 0)
    fmt.Println("This log will never show :(")
}()

time.Sleep(time.Second)
conn.Close()
```

Now I expected the `BRPOP` call to be interrupted with some error as the connection closes but nothing happened.

I suspect this happens only with connection pools where the real underlying connection is not cut for some reason.

If I do:

```
redisPool.Close()
```

Instead, then I observe the behavior I expect. That is, the call is interrupted and fails with an error. Sadly, this isn't an option for me as closing the whole pool is way overkill.

Is there a way to make `Conn.Close()` produce my desired behavior ?

I am willing to make a PR but I'll need some guidance as to where to start/how to proceed.

**Comment by cl1337** [comment]
how did `redisPool`  created? Did you start a new pool with timeout?

I tried your scenario with following pool without a problem
`
func newPool() (*redis.Pool, error) {
	timeout := time.Millisecond * 100
	pool := &redis.Pool{
		MaxIdle: 20,
		MaxActive: 50,
		IdleTimeout: timeout,
		Wait: false,
		Dial: func() (redis.Conn, error) {
			c, err := redis.Dial(
				"tcp",
				"localhost:6379",
				redis.DialConnectTimeout(timeout),
				redis.DialWriteTimeout(timeout),
				redis.DialReadTimeout(timeout),
			)
			if err != nil {
				return nil, err
			}
			return c, nil
		},
	}

	conn := pool.Get()
	defer conn.Close()
	_, err := conn.Do("PING")
	if err != nil {
		return nil, err
	}
	return pool, nil
}
`

**Comment by ereOn** [comment]
@cl1337 I believe that is not same scenario:

In your case the `Do` call expires because of a timeout and `Close` is only called on scope-exit. In my case, I expect the `Close` call to actually interrupt the `Do` call.

Have you tried my example ?

**Comment by cl1337** [comment]
sorry let me rephrase, in your example I didn't see how `redisPool` implemented, I was using a pool implementation I mentioned above in your example and it worked for me

```
redisPool, _ := newPool() // the new pool I mentioned above

conn := redisPool.Get()
defer conn.Close()

go func() {
    // Blocking call. We block forever (0 timeout).
    redisClient.Do("BRPOP", keyQueue, 0)
    fmt.Println("This log will never show :(")
}()

time.Sleep(time.Second)
conn.Close()
```

**Comment by ereOn** [comment]
@cl1337 

I just tried with the following sample:

```
package main

import (
	"fmt"
	"time"

	"github.com/garyburd/redigo/redis"
)

func newPool() (*redis.Pool, error) {
	timeout := time.Second * 10
	pool := &redis.Pool{
		MaxIdle:     20,
		MaxActive:   50,
		IdleTimeout: timeout,
		Wait:        false,
		Dial: func() (redis.Conn, error) {
			c, err := redis.Dial(
				"tcp",
				"127.0.0.1:6379",
				redis.DialConnectTimeout(timeout),
				redis.DialWriteTimeout(timeout),
				redis.DialReadTimeout(timeout),
			)
			if err != nil {
				return nil, err
			}
			return c, nil
		},
	}

	conn := pool.Get()
	defer conn.Close()
	_, err := conn.Do("PING")
	if err != nil {
		return nil, err
	}
	return pool, nil
}

func main() {
	redisPool, err := newPool() // the new pool I mentioned above

	if err != nil {
		panic(err)
	}

	wait := make(chan struct{})
	fmt.Println("Entering")

	conn := redisPool.Get()

	go func() {
		// Blocking call. We block forever (0 timeout).
		_, err := conn.Do("BRPOP", "foo", 0)
		fmt.Println("This log will never show :(", err)
		close(wait)
	}()

	time.Sleep(time.Second)
	fmt.Println("Leaving")
	defer conn.Close()
	<-wait
}
```

And the problem appears to be present: I have to wait for the timeout to reach its value before the goroutine unblocks, even though the connection is cut.

**Comment by garyburd** [comment]
Concurrent calls to `Close` and `Do` are [not supported](https://godoc.org/github.com/garyburd/redigo/redis#hdr-Concurrency). 

I suggest calling BRPOP in a loop with a non-zero timeout.

	const timeout = 1
	done := make(chan struct{})
	go func() {
		conn := redisPool.Get()
		defer conn.Close()
		for {
			select {
			case <-done:
				return
			default:
				resp, err := redis.ByteSlices(c.Do("BRPOP", key, timeout))
				if err == redis.ErrNil {
                    // timeout
					continue
				}
				if err != nil {
					log.Fatal(err) // handle error in some way
				}
				// process result of BRPOP
			}
		}
	}()
	time.Sleep(time.Second)
	close(done)

**Comment by garyburd** [comment]
The Pool Close method does nothing with connections in use.  Are you sure that calling `redisPool.Close()` instead of `conn.Close()` has the behavior you expect?

**Comment by ereOn** [comment]
@garyburd Yep. Closing the pool apparently causes the underlying connection to close for some reason. It might have something to do with the pool size perhaps ?

I did not set any timeout of any kind though.

**Comment by garyburd** [comment]
The pool Close method [closes idle connections](https://github.com/garyburd/redigo/blob/0d253a66e6e1349f4581d6d2b300ee434ee2da9f/redis/pool.go#L203-L205) only.  This is a digression from the main issue.  You should call BRPOP with a timeout.

**Comment by ereOn** [comment]
@garyburd Calling with a timeout doesn't solve the problem though: even with a one second timeout and polling, I still have to wait 1 second for each such goroutine to exit.

This is a no-go with the kind of load we have to deal with.

**Comment by garyburd** [comment]
If closing and redialing connections will work for you, then do this:

       c, err := pool.Dial()
       if err != nil {
           / /handle error
       }
       ... do stuff ...
      c.Close()

This code gets a connection that is not managed by the pool.
